### PR TITLE
Fix: Investigate initializer errors

### PR DESCRIPTION
**Pull request by AsimRibo**
## :page_facing_up: Context

When user wouldn't include all tools into `build.gradle` file, it would cause crash "WorkManager is already initialized". It seems problem was in how `AndroidManifests` were merged.
I am adding Antun as optional reviewer since he was there when crashes happened 😄 .

## :pencil: Changes

By removing 2 startup providers and merging them under one with separate meta-data, crashes are gone. Same provider had to be put inside of no-op module as well, otherwise it would cause NPE. Other modules didn't have to be touched.

### CODE REVIEW

File: sentinel/src/main/AndroidManifest.xml
Comment (antunflas) [inline sentinel/src/main/AndroidManifest.xml:121]: Why is Sentinel removing the work manager initializer? Does it even use anything from the work manager?

I'm worried about this because if the app add sentinel they would "hiddenly" remove the initialization of the work manager. Everything would be fine for the apps using [custom work manager initialization](https://developer.android.com/develop/background-work/background-tasks/persistent/configuration/custom-configuration#remove-default), but for the apps using the default one, work manager wouldn't be initialized anymore.

I'd be calmer if we would remove this code completely, therefore not removing the work manager initialization, and therefore leaving the work manager setup as it is defined in the apps.

Comment (AsimRibo) [reply sentinel/src/main/AndroidManifest.xml:121]: I will look into that option as well. It should work.

Comment (AsimRibo) [reply sentinel/src/main/AndroidManifest.xml:121]: Just to double check. You are suggesting to remove this whole part?
```
<meta-data
                android:name="androidx.work.WorkManagerInitializer"
                android:value="androidx.startup"
                tools:node="remove" />
```

Do you then think we should add this to our README as a reminder for users that they will have to do it in their projects?
I will leave the decision to Karlo as he is the owner.

Comment (antunflas) [reply sentinel/src/main/AndroidManifest.xml:121]: Yes, I'm suggesting that, but not in case users would need to add this to their projects only because they have Sentinel.
I'm suggesting removing this if it is redundant.
The only thing when users should add this is in case they use a work manager in their project.

Does Sentinel even use a work manager? What for?

Comment (AsimRibo) [reply sentinel/src/main/AndroidManifest.xml:121]: I think only usage is in `SentinelWorkManager` class. It is used for certificates checking. Interestingly enough, this is the second time certificates causing problems 😄 .

Comment (antunflas) [reply sentinel/src/main/AndroidManifest.xml:121]: And I think that is a problem. Since WorkManager can be initialized only once (next time it throws), any application using Work Manager would fail to supply their own workers and initialize the WorkManager. That means, the application dev has to choose, either to use WorkManager and no Sentinel, or vice-versa.

That's the problem we should investigate further and see how to avoid the clash.

**Comment by KCeh** [comment]
I made a fix based on https://developer.squareup.com/blog/workmanager-for-background-work-in-libraries/

I changed 2 things:

- Sentinel will initialize `WorkManager` **only** if it hasn't been initialized
- In Sentinel now we invoke `DelegateWorker` (instead of `CertificateCheckWorker`) to do work. `DelegateWorker` will create an appropriate worker and delegate work to it. The only additional trick is to remember `WorkerFactory` for our `CertificateCheckWorker` inside of `DelegateWorker`. Why do we need to do that?
Because `CertificateCheckWorker` (and related factory) have custom dependencies and `WorkManager` won't be able to create them without setting Configuration (which we don't set up if app initializes `WorkManager`)

How I tested:

- I added Sentinel to New Project Templated (only main Sentinel lib without extra tools). The old version of Sentinel crashed the app, while the new one (with this PR) doesn't.

- I have also tested the sample app and added a log to see if `doWork` is invoked properly in `CertificateCheckWorker`. Also, I managed to get notifications from Sentinel regarding the invalid certificate (notifications have to be enabled!). So I would say it works!

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt
Comment (AsimRibo) [inline sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:61]: Should we include thread annotation here?

**Review by AsimRibo — COMMENTED** [review]
Looks good 🔥 .
Do we need to have 2 different `provider` elements in `AndroidManifest`? I think `meta data` can still be under one `provider` element.
Also, I can't approve this because I am still author 😂 . Should we ping Flas or someone?

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt
Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:61]: Will look into this

**Comment by KCeh** [comment]
> Do we need to have 2 different `provider` elements in `AndroidManifest`? I think `meta data` can still be under one `provider` element

I haven't look into that to be honest, I will explore

>  Also, I can't approve this because I am still author 😂 . Should we ping Flas or someone?

Well, I can approve PR for myself? But, yes, if someone else can take a look why not?

**Comment by KCeh** [comment]
Are we good with this PR?

**Comment by AsimRibo** [comment]
> Are we good with this PR?

From my side good. I pinged Antun today but I believe he is busy.

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt
Comment (antunflas) [inline sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:50]: Why do we return an "empty" future instead of throwing an error? Shouldn't throwing an error be safer?

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt
Comment (antunflas) [inline sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt:68]: I can't connect pieces, why is this different than before? I mean in both cases, we pass the worker we want to run, shouldn't then even the previous line work fine?

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt
Comment (antunflas) [inline sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt:47]: What WorkerFactory do we inject here?

Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt:47]: `SentinelWorkerFactory` which knows how to create `CertificateCheckWorker`
But, now actually you are mentioning it, this shouldn't be needed anymore

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt
Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:50]: Well, it is not "empty". It is failure with errorData. We need to return `ListenableFuture<Result>`

File: sentinel/src/main/kotlin/com/infinum/sentinel/di/component/DomainComponent.kt
Comment (antunflas) [inline sentinel/src/main/kotlin/com/infinum/sentinel/di/component/DomainComponent.kt:167]: Let's avoid the put inside of a provides method because it will be executed only when someone injects SentinelWorkerFactory. That can be either very soon or sometime later - and it should be set immediately because anyone using the work manager should be able to use it without other dependencies.

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt
Comment (antunflas) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:50]: Yeah, but I'm exploring less "silent" ways of this failing. Realistically, it could happen only if we forget to register a worker factory, but that's something we shouldn't allow to end up in a release. I'm trying to figure out how we could make sure to not fail fast and loud :D

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt
Comment (antunflas) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt:47]: Let's remove it then. I'd remove the logging as well, or set it conditionally only for debug 🤷‍♂️

File: sentinel/src/main/kotlin/com/infinum/sentinel/di/component/DomainComponent.kt
Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/di/component/DomainComponent.kt:167]: Good point 👍

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt
Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:50]: Do you think just exception will do the job?

Comment (antunflas) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:50]: I'm open for discussions, but an exception (if it would crash the app) would definitely be more noticeable.

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt
Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt:47]: If I remove debugging it seems we don't need to call `WorkManager.initialize` and then we can also `WorkManagerInitializer` from Manifest

Comment (antunflas) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/SentinelWorkManager.kt:47]: Does this also work when the host application doesn't initialize the work manager?

File: sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt
Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:50]: If we throw exception in this place it will just produce logs, similar to Result.Failure. Because it is running on background thread 
```
Work [ id=42358918-83db-4c07-847b-6020296c3d41, tags={ com.infinum.sentinel.ui.certificates.observer.DelegateWorker } ] failed because it threw an exception/error
                                                                                                    java.util.concurrent.ExecutionException: java.lang.IllegalStateException: No delegateWorker available for com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker with workerClassName of com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker. Is the DelegateWorker.workerFactories populated correctly?
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:515)
                                                                                                    	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
                                                                                                    	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
                                                                                                    	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
                                                                                                    Caused by: java.lang.IllegalStateException: No delegateWorker available for com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker with workerClassName of com.infinum.sentinel.ui.certificates.observer.CertificateCheckWorker. Is the RouterWorker.workerFactories populated correctly?
                                                                                                    	at com.infinum.sentinel.ui.certificates.observer.DelegateWorker.startWork(DelegateWorker.kt:40)
                                                                                                    	at androidx.work.impl.WorkerWrapper$1.run(WorkerWrapper.java:302)
                                                                                                    	at android.os.Handler.handleCallback(Handler.java:942)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:99)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:201)
                                                                                                    	at android.os.Looper.loop(Looper.java:288)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:7872)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
2024-07-17 16:57:47.222 15350-15401 WM-WorkerWrapper        com.infinum.sentinel.sample.debug    I  Worker result FAILURE for Work [ id=42358918-83db-4c07-847b-6020296c3d41, tags={ com.infinum.sentinel.ui.certificates.observer.DelegateWorker } ] 
```

But error is defiantly more noticeable in logs

Comment (KCeh) [reply sentinel/src/main/kotlin/com/infinum/sentinel/ui/certificates/observer/DelegateWorker.kt:50]: If things go 💥 at least people will see errors and open issues 😅

**Comment by KCeh** [comment]
All points should be improved upon now

@AsimRibo recheck one more time. Just don't forget to bump version in `config.gradle `and `libs.toml` publish to mavenLocal and used that new version. Because tools have references to published version and if you don't update that before test it could cause errors.

File: sentinel/src/main/kotlin/com/infinum/sentinel/di/LibraryComponents.kt
Comment (KCeh) [inline sentinel/src/main/kotlin/com/infinum/sentinel/di/LibraryComponents.kt:45]: Needs to be called here, else tools won't be visible in UI

Comment (antunflas) [reply sentinel/src/main/kotlin/com/infinum/sentinel/di/LibraryComponents.kt:45]: Why can't it be called from initialize method? Why the tools won't be visible in the UI?

Comment (antunflas) [reply sentinel/src/main/kotlin/com/infinum/sentinel/di/LibraryComponents.kt:45]: Okay, thanks for the explanation 👍

**Review by antunflas — APPROVED** [review]

**Comment by AsimRibo** [comment]
> All points should be improved upon now
> 
> @AsimRibo recheck one more time. Just don't forget to bump version in `config.gradle `and `libs.toml` publish to mavenLocal and used that new version. Because tools have references to published version and if you don't update that before test it could cause errors.

Looking good amigo.

**Pull request merged**
